internal/pkg/interp/interpreter.go

```diff
@@ -0,0 +1,425 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package interp defines an analyzer that performs taint analysis on
+// functions by interpreting their ssa code.
+package interp
+
+import (
+	"go/token"
+	"go/types"
+	"reflect"
+
+	"github.com/google/go-flow-levee/internal/pkg/config"
+	"github.com/google/go-flow-levee/internal/pkg/fieldpropagator"
+	"github.com/google/go-flow-levee/internal/pkg/utils"
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/buildssa"
+	"golang.org/x/tools/go/pointer"
+	"golang.org/x/tools/go/ssa"
+)
+
+// A Result contains the information necessary to report that
+// a Source has reached a Sink. Currently this is just the
+// source code positions of the Source and Sink.
+type Result struct {
+	SourcePos, SinkPos token.Pos
+}
+
+// A blocksKey is used to determine whether the current block has already
+// been visited from the pred(ecessor) block.
+type blocksKey struct {
+	current, pred *ssa.BasicBlock
+}
+
+// A val represents a Value that is tainted by a source.
+type val struct {
+	source ssa.Value
+}
+
+// A ptr represents a Value that points to another Value. This implies that
+// taint needs to be propagated to and from the pointed to value.
+type ptr struct {
+	pointsTo ssa.Value
+}
+
+// An interpreterState is a mapping from ssa Values to either a val or a ptr.
+type interpreterState map[ssa.Value]interface{}
+
+// An interpreter is used to perform taint analysis on a function by
+// stepping through its instructions. When a source reaches a sink,
+// a result is added to its slice of Results.
+type interpreter struct {
+	conf             *config.Config
+	fieldpropagators fieldpropagator.ResultType
+	results          []Result
+}
+
+// ResultType is a slice of Results.
+type ResultType = []Result
+
+// Analyzer is the analysis.Analyzer for the interp package.
+var Analyzer = &analysis.Analyzer{
+	Name: "interp",
+	Doc:  "Performs taint propagation analysis by stepping through SSA instructions.",
+	Run:  run,
+	Requires: []*analysis.Analyzer{
+		buildssa.Analyzer, fieldpropagator.Analyzer,
+	},
+	ResultType: reflect.TypeOf(new(ResultType)).Elem(),
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	conf, err := config.ReadConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	ssaInput := pass.ResultOf[buildssa.Analyzer].(*buildssa.SSA)
+	fieldPropagators := pass.ResultOf[fieldpropagator.Analyzer].(fieldpropagator.ResultType)
+
+	results := []Result{}
+	for _, f := range ssaInput.SrcFuncs {
+		results = append(results, analyzeFunction(conf, fieldPropagators, f)...)
+	}
+
+	return results, nil
+}
+
+// analyzeFunction performs taint analysis on the given function and
+// returns a slice of Results.
+func analyzeFunction(conf *config.Config, fieldpropagators fieldpropagator.ResultType, f *ssa.Function) []Result {
+	// nothing to analyze
```

Comment:
https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

here and other places.

Reply:
That link seems to say that comments for *declarations* should be full sentences, which I believe they are (e.g. in this snippet I think the doc comment for `analyzeFunction` is in good form). I don't have the link, but I remember reading elsewhere that if a comment is just a couple words and not a complete sentence, then it is ok to omit capitalizing the first word + ending with a period.

Reply:
I re-read internal google style guide on this. You are correct that this does not need to be capitalized or punctuated. Thanks!